python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+
+    @property
+    def instances(self) -> Dict[CloudInstanceId, CloudInstance]:
+        return copy.deepcopy(self._cached_instances)
+
+    @staticmethod
+    def _get_workers_groups_with_deletes(
+        ray_cluster_spec: Dict[str, Any], node_set: Set[CloudInstanceId]
+    ) -> Tuple[Set[NodeType], Set[NodeType]]:
+        """
+        Gets the worker groups that have pending deletes and the worker groups that
+        have finished deletes.
+
+        Returns:
+            worker_groups_with_pending_deletes: The worker groups that have pending
+                deletes.
+            worker_groups_with_finished_deletes: The worker groups that have finished
+                deletes.
+        """
+
+        worker_groups_with_pending_deletes = set()
+        worker_groups_with_finished_deletes = set()
+
+        worker_groups = ray_cluster_spec["spec"].get("workerGroupSpecs", [])
+        for worker_group in worker_groups:
+            workersToDelete = worker_group.get("scaleStrategy", {}).get(
+                "workersToDelete", []
+            )
+            if not workersToDelete:
+                # No workers to delete in this group.
+                continue
+
+            node_type = worker_group["groupName"]
+            for worker in workersToDelete:
+                if worker in node_set:
+                    worker_groups_with_pending_deletes.add(node_type)
+                    break
+
+            # All to delete in the worker spec have been deleted.
+            worker_groups_with_finished_deletes.add(node_type)
+
+        return worker_groups_with_pending_deletes, worker_groups_with_finished_deletes
+
+    def _fetch_instances(self) -> Dict[CloudInstanceId, CloudInstance]:
+        """
+        Fetches the pods from the Kuberay API server and convert them to Ray
+        CloudInstance.
+
+        Returns:
+            A dict of CloudInstanceId to CloudInstance.
+        """
+        # Get the pods resource version.
+        # Specifying a resource version in list requests is important for scalability:
+        # https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list
+        resource_version = self._get_pods_resource_version()
+        if resource_version:
+            logger.info(
+                f"Listing pods for RayCluster {self._cluster_name}"
+                f" in namespace {self._namespace}"
+                f" at pods resource version >= {resource_version}."
+            )
+
+        # Filter pods by cluster_name.
+        label_selector = requests.utils.quote(f"ray.io/cluster={self._cluster_name}")
+
+        resource_path = f"pods?labelSelector={label_selector}"
+        if resource_version:
+            resource_path += (
+                f"&resourceVersion={resource_version}"
+                + "&resourceVersionMatch=NotOlderThan"
+            )
+
+        pod_list = self._get(resource_path)
+        fetched_resource_version = pod_list["metadata"]["resourceVersion"]
```

Comment:
looks like `fetched_resource_version` was merely logged. I wasn't sure why we used it in the first place. cc @DmitriGekhtman

Reply:
The `fetched_resource_version` is at least the supplied `resource_version` in the list request.
It is logged simply for observability to reasons, to give an indication of the potential staleness of the returned pod list data.

Although the K8s API forbids making decisions based on resource version comparisons, it could be useful (at least for maintainers of this code, at least for the ones who wrote this code...) to have a sense of whether this fetched resource version is _at least increasing_ across autoscaler operations.

Feel free to drop the bit of code around the fetched_resource_version if you wish; I essentially added the log line _for myself_.

Providing a bit more context:
The resource version is included in the request with a "not older than parameter" to improve scalability (without this parameter, the requests could be harmful to the health of the K8s cluster, especially when many Ray clusters are running).
Previously, I tried to extract a _recent_ resource version by writing a patch to head pod and then reading the head pod object back; however users complained that the illicit write to the head pod was screwing with their CI/CD setups. So we no longer write to the head pod and just keep reusing the resource version at which the head pod was last updated (the head pod was potentially last updated when it started running), just accept that the read might be a little stale, and hope that the read is at least fresh enough to reflect changes since the last autoscaler update.
In theory, **this is sketch**. In practice, it seems that this hasn't caused any issues for anyone...yet...that we've heard of.

Reply:
Thanks for the info! This is very helpful for learning the context. 

Doesn't look like we are using `fetched_resource_version` other than logging? I am fine with keeping things as it's now and we could revisit it.